File: amazon-intent-os-main/src/lib/scoring/rankingEngine.ts
// src/lib/scoring/rankingEngine.ts
import { ProductItem, RankedItem, UserProfile } from "../../types/inventory";

export const rankAndFlagAlternatives = (
  inventory: ProductItem[],
  userProfile: UserProfile,
  isSearch: boolean = false
): RankedItem[] => {
  
 let scoredItems: RankedItem[] = inventory.map(item => {
    const currentPrice = item.base_price * item.surge_multiplier;
    let baseScore = (item.purchase_frequency_rank * 0.8) - (item.eta_mins * 3) - (currentPrice * 0.05);
    let score = Math.max(0, baseScore);
    
    // Personalization Boost: If the user loves this brand, boost its score heavily
    if (userProfile.preferred_brands.includes(item.brand)) {
      score += 75; 
    }
    
    return { 
      ...item, 
      score, 
      is_alternative: false 
    };
  });

  // 2. Sort by highest score to get the absolute best options
  scoredItems.sort((a, b) => b.score - a.score);

  // 3. Process stock checks and dynamic alternative flags
  let finalTopItems: RankedItem[] = [];
  let isAlternativeTriggered = false;
  let originalBrand = "";

  for (const currentItem of scoredItems) {
    // If an item is out of stock, we skip it.
    // If it's a direct search AND the absolute best item was out of stock, flag for an alternative.
    if (currentItem.stock_count === 0) {
      if (isSearch && finalTopItems.length === 0) {
        isAlternativeTriggered = true;
        originalBrand = currentItem.brand;
      }
      continue; 
    }

    let processedItem = { ...currentItem };

    // Apply the Alternative Banner to the NEXT best item
    if (isSearch && isAlternativeTriggered && finalTopItems.length === 0) {
      processedItem.is_alternative = true;
      processedItem.alternative_message = `${originalBrand} is out. Suggested Substitute: ${processedItem.brand}`;
    }

    finalTopItems.push(processedItem);
    
    // We strictly return only the Top 3 items for the Showdown Drawer if it's a direct search
    if (isSearch && finalTopItems.length === 3) break; 
  }

  return finalTopItems;
};

File: cart-oracle-lakshan/backend/ranking_engine.py
# ranking_engine.py - Ported from Project B's rankingEngine.ts
# Scores and ranks products, flags alternatives when items are out of stock

from seed_data_extended import EXTENDED_INVENTORY, USER_PROFILE


def rank_and_flag_alternatives(inventory: list, user_profile: dict, is_search: bool = False) -> list:
    """
    Rank products using algorithmic scoring:
    - purchase_frequency_rank contributes positively
    - eta_mins penalizes slow delivery
    - current price (base * surge) penalizes expensive items
    - Brand preference gives a large boost
    
    Also flags alternatives when top item is out of stock.
    """
    scored_items = []

    for item in inventory:
        current_price = item["base_price"] * item["surge_multiplier"]
        base_score = (item["purchase_frequency_rank"] * 0.8) - (item["eta_mins"] * 3) - (current_price * 0.05)
        score = max(0, base_score)

        # Personalization boost
        if item["brand"] in user_profile.get("preferred_brands", []):
            score += 75

        scored_items.append({**item, "score": score, "is_alternative": False})

    # Sort by highest score
    scored_items.sort(key=lambda x: x["score"], reverse=True)

    # Process stock checks and dynamic alternative flags
    final_items = []
    is_alternative_triggered = False
    original_brand = ""

    for current_item in scored_items:
        if current_item["stock_count"] == 0:
            if is_search and len(final_items) == 0:
                is_alternative_triggered = True
                original_brand = current_item["brand"]
            continue

        processed_item = {**current_item}

        # Apply alternative banner to next best item
        if is_search and is_alternative_triggered and len(final_items) == 0:
            processed_item["is_alternative"] = True
            processed_item["alternative_message"] = f"{original_brand} is out. Suggested Substitute: {processed_item['brand']}"

        final_items.append(processed_item)

        # Return only top 3 for direct search
        if is_search and len(final_items) == 3:
            break

    return final_items


def fetch_inventory_by_category(category: str, macro_crisis: str = None) -> list:
    """Filter extended inventory by category or crisis type."""
    normalized_category = category.lower().strip().replace(" ", "_") if category else ""
    normalized_crisis = macro_crisis.lower().strip().replace(" ", "_") if macro_crisis else ""

    results = [
        item for item in EXTENDED_INVENTORY
        if (normalized_category and item["category"].lower() == normalized_category)
        or (normalized_crisis and item["macro_crisis"].lower() == normalized_crisis)
    ]
    return results


def get_matched_products(target_category: str, macro_crisis: str, is_search: bool = False) -> list:
    """Full pipeline: fetch inventory -> rank -> flag alternatives."""
    inventory = fetch_inventory_by_category(target_category, macro_crisis)
    ranked = rank_and_flag_alternatives(inventory, USER_PROFILE, is_search)
    return ranked
